sessionmanager/sessionmanager.go

```diff
@@ -0,0 +1,65 @@
+package sessionmanager
+
+import (
+	"sync"
+
+	exchange "github.com/ipfs/go-ipfs-exchange-interface"
+)
+
+type SessionManager interface {
+	GetNextSessionID() uint64
+	IterateSessions(iterate iterateSessionFunc)
+	AddSession(session exchange.Fetcher)
+	RemoveSession(session exchange.Fetcher)
+}
+
+type sessionManager struct {
+	// Sessions
+	sessions []exchange.Fetcher
+	sessLk   sync.Mutex
+
+	sessID   uint64
+	sessIDLk sync.Mutex
+}
+
+func New() SessionManager {
+	return &sessionManager{}
+}
+
+func (sm *sessionManager) AddSession(session exchange.Fetcher) {
+	sm.sessLk.Lock()
+	defer sm.sessLk.Unlock()
```

Comment:
For simple locking scenarios the performance overhead of `defer` outweighs the code simplicity. I suggest calling `Unlock` directly after the `append` operation.